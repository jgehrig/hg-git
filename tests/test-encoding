#!/bin/sh

# -*- coding: utf-8 -*-

# Fails for some reason, need to investigate
# "$TESTDIR/hghave" git || exit 80

# bail if the user does not have dulwich
python -c 'import dulwich, dulwich.repo' || exit 80

# bail early if the user is already running git-daemon
echo hi | nc localhost 9418 2>/dev/null && exit 80

echo "[extensions]" >> $HGRCPATH
echo "hggit=$(echo $(dirname $(dirname $0)))/hggit" >> $HGRCPATH
echo 'hgext.graphlog =' >> $HGRCPATH
echo 'hgext.bookmarks =' >> $HGRCPATH

GIT_AUTHOR_NAME='test'; export GIT_AUTHOR_NAME
GIT_AUTHOR_EMAIL='test@example.org'; export GIT_AUTHOR_EMAIL
GIT_AUTHOR_DATE="2007-01-01 00:00:00 +0000"; export GIT_AUTHOR_DATE
GIT_COMMITTER_NAME="$GIT_AUTHOR_NAME"; export GIT_COMMITTER_NAME
GIT_COMMITTER_EMAIL="$GIT_AUTHOR_EMAIL"; export GIT_COMMITTER_EMAIL
GIT_COMMITTER_DATE="$GIT_AUTHOR_DATE"; export GIT_COMMITTER_DATE

count=10
commit()
{
    GIT_AUTHOR_DATE="2007-01-01 00:00:$count +0000"
    GIT_COMMITTER_DATE="$GIT_AUTHOR_DATE"
    git commit "$@" >/dev/null || echo "git commit error"
    count=`expr $count + 1`
}

mkdir gitrepo
cd gitrepo
git init | python -c "import sys; print sys.stdin.read().replace('$(dirname $(pwd))/', '')"

# utf-8 encoded commit message
echo alpha > alpha
git add alpha
commit -m 'add älphà'

. $TESTDIR/latin-1-encoding

cd ..
mkdir gitrepo2
cd gitrepo2
git init --bare | python -c "import sys; print sys.stdin.read().replace('$(dirname $(pwd))/', '')"

# dulwich does not presently support local git repos, workaround
cd ..
git daemon --base-path="$(pwd)"\
 --listen=localhost\
 --export-all\
 --pid-file="$DAEMON_PIDS" \
 --detach --reuseaddr \
 --enable=receive-pack

hg clone git://localhost/gitrepo hgrepo | grep -v '^updating'
cd hgrepo

HGENCODING=utf-8 hg log --graph --debug | grep -v ': *master'

hg gclear
hg push git://localhost/gitrepo2

cd ../gitrepo2
git log --pretty=medium

cd ..
