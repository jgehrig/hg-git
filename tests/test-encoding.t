# -*- coding: utf-8 -*-

Load commonly used test logic
  $ . "$TESTDIR/testutil"

  $ git init gitrepo
  Initialized empty Git repository in $TESTTMP/gitrepo/.git/
  $ cd gitrepo

utf-8 encoded commit message
  $ echo alpha > alpha
  $ git add alpha
  $ fn_git_commit -m 'add älphà'

Create some commits using latin1 encoding
The warning message changed in Git 1.8.0
  $ . $TESTDIR/latin-1-encoding
  Warning: commit message (did|does) not conform to UTF-8. (re)
  You may want to amend it after fixing the message, or set the config
  variable i18n.commitencoding to the encoding your project uses.
  Warning: commit message (did|does) not conform to UTF-8. (re)
  You may want to amend it after fixing the message, or set the config
  variable i18n.commitencoding to the encoding your project uses.

  $ cd ..
  $ git init --bare gitrepo2
  Initialized empty Git repository in $TESTTMP/gitrepo2/

  $ hg clone gitrepo hgrepo | grep -v '^updating'
  importing git objects into hg
  4 files updated, 0 files merged, 0 files removed, 0 files unresolved
  $ cd hgrepo

  $ HGENCODING=utf-8 hg log --graph --debug | grep -v 'phase:' | grep -v ': *author=' | grep -v ': *message='
  @  changeset:   3:b8a0ac52f339ccd6d5729508bac4aee6e8b489d8
  |  bookmark:    master
  |  tag:         default/master
  |  tag:         tip
  |  parent:      2:8bc4d64940260d4a1e70b54c099d3a76c83ff41e
  |  parent:      -1:0000000000000000000000000000000000000000
  |  manifest:    3:ea49f93388380ead5601c8fcbfa187516e7c2ed8
  |  user:        tést èncödîng <test@example.org>
  |  date:        Mon Jan 01 00:00:13 2007 +0000
  |  files+:      delta
  |  extra:       branch=default
  |  extra:       committer=test <test@example.org> 1167609613 0
  |  extra:       encoding=latin-1
  |  extra:       hg-git-rename-source=git
  |  description:
  |  add d\xc3\xa9lt\xc3\xa0 (esc)
  |
  |
  o  changeset:   2:8bc4d64940260d4a1e70b54c099d3a76c83ff41e
  |  parent:      1:f35a3100b78e57a0f5e4589a438f952a14b26ade
  |  parent:      -1:0000000000000000000000000000000000000000
  |  manifest:    2:f580e7da3673c137370da2b931a1dee83590d7b4
  |  user:        t\xc3\xa9st \xc3\xa8nc\xc3\xb6d\xc3\xaeng <test@example.org> (esc)
  |  date:        Mon Jan 01 00:00:12 2007 +0000
  |  files+:      gamma
  |  extra:       branch=default
  |  extra:       committer=test <test@example.org> 1167609612 0
  |  extra:       hg-git-rename-source=git
  |  description:
  |  add g\xc3\xa4mm\xc3\xa2 (esc)
  |
  |
  o  changeset:   1:f35a3100b78e57a0f5e4589a438f952a14b26ade
  |  parent:      0:87cd29b67a9159eec3b5495b0496ef717b2769f5
  |  parent:      -1:0000000000000000000000000000000000000000
  |  manifest:    1:f0bd6fbafbaebe4bb59c35108428f6fce152431d
  |  user:        t\xc3\xa9st \xc3\xa8nc\xc3\xb6d\xc3\xaeng <test@example.org> (esc)
  |  date:        Mon Jan 01 00:00:11 2007 +0000
  |  files+:      beta
  |  extra:       branch=default
  |  extra:       committer=test <test@example.org> 1167609611 0
  |  extra:       hg-git-rename-source=git
  |  description:
  |  add beta
  |
  |
  o  changeset:   0:87cd29b67a9159eec3b5495b0496ef717b2769f5
     parent:      -1:0000000000000000000000000000000000000000
     parent:      -1:0000000000000000000000000000000000000000
     manifest:    0:8b8a0e87dfd7a0706c0524afa8ba67e20544cbf0
     user:        test <test@example.org>
     date:        Mon Jan 01 00:00:10 2007 +0000
     files+:      alpha
     extra:       branch=default
     extra:       hg-git-rename-source=git
     description:
     add \xc3\xa4lph\xc3\xa0 (esc)
  
  

  $ hg gclear
  clearing out the git cache data
  $ hg push ../gitrepo2
  pushing to ../gitrepo2
  searching for changes
  adding objects
  added 4 commits with 4 trees and 4 blobs

  $ cd ..
  $ git --git-dir=gitrepo2 log --pretty=medium
  commit e85fef6b515d555e45124a5dc39a019cf8db9ff0
  Author: t\xe9st \xe8nc\xf6d\xeeng <test@example.org> (esc)
  Date:   Mon Jan 1 00:00:13 2007 +0000
  
      add d\xe9lt\xe0 (esc)
  
  commit bd576458238cbda49ffcfbafef5242e103f1bc24
  Author: * <test@example.org> (glob)
  Date:   Mon Jan 1 00:00:12 2007 +0000
  
      add g*mm* (glob)
  
  commit 7a7e86fc1b24db03109c9fe5da28b352de59ce90
  Author: * <test@example.org> (glob)
  Date:   Mon Jan 1 00:00:11 2007 +0000
  
      add beta
  
  commit 0530b75d8c203e10dc934292a6a4032c6e958a83
  Author: test <test@example.org>
  Date:   Mon Jan 1 00:00:10 2007 +0000
  
      add älphà
