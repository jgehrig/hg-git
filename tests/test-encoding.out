Initialized empty Git repository in gitrepo/.git/

Warning: commit message does not conform to UTF-8.
You may want to amend it after fixing the message, or set the config
variable i18n.commitencoding to the encoding your project uses.
Warning: commit message does not conform to UTF-8.
You may want to amend it after fixing the message, or set the config
variable i18n.commitencoding to the encoding your project uses.
Initialized empty Git repository in gitrepo2/

importing git objects into hg
4 files updated, 0 files merged, 0 files removed, 0 files unresolved
@  changeset:   3:8549ee7fe0801b2dafc06047ca6f66d36da709f5
|  tag:         default/master
|  tag:         tip
|  parent:      2:0422fbb4ec39fb69e87b94a3874ac890333de11a
|  parent:      -1:0000000000000000000000000000000000000000
|  manifest:    3:ea49f93388380ead5601c8fcbfa187516e7c2ed8
|  user:        t√©st √®nc√∂d√Æng <test@example.org>
|  date:        Mon Jan 01 00:00:13 2007 +0000
|  files+:      delta
|  extra:       author=$ \x90\x01\x01\xe9\x91\x03\x03\x01\xe8\x91\x08\x02\x01\xf6\x91\x0c\x01\x01\xee\x91\x0f\x15
|  extra:       branch=default
|  extra:       committer=test <test@example.org> 1167609613 0
|  extra:       encoding=latin-1
|  extra:       message=\x0c\n\x90\x05\x01\xe9\x91\x07\x02\x01\xe0\x91\x0b\x01
|  description:
|  add d√©lt√†
|
|
o  changeset:   2:0422fbb4ec39fb69e87b94a3874ac890333de11a
|  parent:      1:9f6268bfc9eb3956c5ab8752d7b983b0ffe57115
|  parent:      -1:0000000000000000000000000000000000000000
|  manifest:    2:f580e7da3673c137370da2b931a1dee83590d7b4
|  user:        t√©st √®nc√∂d√Æng <test@example.org>
|  date:        Mon Jan 01 00:00:12 2007 +0000
|  files+:      gamma
|  extra:       author=$ \x90\x01\x01\xe9\x91\x03\x03\x01\xe8\x91\x08\x02\x01\xf6\x91\x0c\x01\x01\xee\x91\x0f\x15
|  extra:       branch=default
|  extra:       committer=test <test@example.org> 1167609612 0
|  extra:       message=\x0c\n\x90\x05\x01\xe4\x91\x07\x02\x01\xe2\x91\x0b\x01
|  description:
|  add g√§mm√¢
|
|
o  changeset:   1:9f6268bfc9eb3956c5ab8752d7b983b0ffe57115
|  parent:      0:bb7d36568d6188ce0de2392246c43f6f213df954
|  parent:      -1:0000000000000000000000000000000000000000
|  manifest:    1:f0bd6fbafbaebe4bb59c35108428f6fce152431d
|  user:        t√©st √®nc√∂d√Æng <test@example.org>
|  date:        Mon Jan 01 00:00:11 2007 +0000
|  files+:      beta
|  extra:       author=$ \x90\x01\x01\xe9\x91\x03\x03\x01\xe8\x91\x08\x02\x01\xf6\x91\x0c\x01\x01\xee\x91\x0f\x15
|  extra:       branch=default
|  extra:       committer=test <test@example.org> 1167609611 0
|  description:
|  add beta
|
|
o  changeset:   0:bb7d36568d6188ce0de2392246c43f6f213df954
   parent:      -1:0000000000000000000000000000000000000000
   parent:      -1:0000000000000000000000000000000000000000
   manifest:    0:8b8a0e87dfd7a0706c0524afa8ba67e20544cbf0
   user:        test <test@example.org>
   date:        Mon Jan 01 00:00:10 2007 +0000
   files+:      alpha
   extra:       branch=default
   description:
   add √§lph√†


clearing out the git cache data
pushing to git://localhost/gitrepo2
exporting hg objects to git
creating and sending data
commit da0edb01d4f3d1abf08b1be298379b0b2960e680
Author: tÈst ËncˆdÓng <test@example.org>
Date:   Mon Jan 1 00:00:13 2007 +0000

    add dÈlt‡

commit 2372b6c8f1b91f2db8ae5eb0f9e0427c318b449c
Author: tÈst ËncˆdÓng <test@example.org>
Date:   Mon Jan 1 00:00:12 2007 +0000

    add g‰mm‚

commit 9ef7f6dcffe643b89ba63f3323621b9a923e4802
Author: tÈst ËncˆdÓng <test@example.org>
Date:   Mon Jan 1 00:00:11 2007 +0000

    add beta

commit 0530b75d8c203e10dc934292a6a4032c6e958a83
Author: test <test@example.org>
Date:   Mon Jan 1 00:00:10 2007 +0000

    add √§lph√†
